llama_stack/providers/remote/inference/nutanix/nutanix.py

```diff
@@ -0,0 +1,147 @@
+# Copyright (c) Nutanix, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the terms described in the LICENSE file in
+# the root directory of this source tree.
+
+from typing import AsyncGenerator
+
+from openai import OpenAI
+
+from llama_models.llama3.api.chat_format import ChatFormat
+
+from llama_models.llama3.api.datatypes import Message
+from llama_models.llama3.api.tokenizer import Tokenizer
+
+from llama_stack.apis.inference import *  # noqa: F403
+from llama_stack.providers.utils.inference.model_registry import (
+    build_model_alias,
+    ModelRegistryHelper,    
+) 
+from llama_stack.providers.utils.inference.openai_compat import (
+    get_sampling_options,
+    process_chat_completion_response,
+    process_chat_completion_stream_response,
+)
+from llama_stack.providers.utils.inference.prompt_adapter import (
+    chat_completion_request_to_messages,
+)
+
+from .config import NutanixImplConfig
+
+
+MODEL_ALIASES = [
+    build_model_alias(
+        "vllm-llama-3-1",
+        CoreModelId.llama3_1_8b_instruct.value,
+    ),
+]
+
+
+class NutanixInferenceAdapter(ModelRegistryHelper, Inference):
```

Comment:
@mattf Yes, I think we need to start consolidating more on the code side. 

we have some tests now but we also need to put down some more requirements of when a new inference provider comes in. here are some things we are thinking about:

- support for structured decoding -- kind of table stakes now 
- proper support for tool calling (either directly or via allow legacy completions API so llama stack can format the prompt) 
- support for vision models 

otherwise we cannot claim to the user that "you can just Llama Stack and pick-and-choose any provider and you will get a consistent experience"

Reply:
Thank you for the feedback. While the code does share similarities with Fireworks and Databricks, there are important differences, and we anticipate adding new features that will further differentiate our implementation from those of other vendors.

I believe it may be more efficient for each vendor to maintain their own Llama Stack adapter. The duplication of code within each adapter, in this context, is manageable and can even be beneficial. Adopting a "[Do Repeat Yourself](https://huggingface.co/blog/transformers-design-philosophy)" approach for these adapters aligns with maintaining clarity and flexibility, especially given the unique requirements and evolution of individual providers.

That said, I’m open to further discussions if there’s a strong case for a shared base class or alternative approach. Let me know your thoughts!

Reply:
@jinan-zhou thank you for the thoughtful argument. i think you're right that abstracting the providers now is too early. i raise the topic only to start a discuss, not to block or slow your valuable contribution.